### ISSUE TITLE
Write module to parse nucmer output and test

### ISSUE THREAD

**Issue by widdowquinn** [pyani-plus/pyani-plus#4]
While developing the snakemake interaction, we need to be confident that our output is correct.

We should use `pytest` to manage this process (so a `tests` directory etc. wil be required) and we will need to parse output/compare it to the expected output (see #1)

**Comment by widdowquinn** [comment]
NOTE: This is on branch `issue_4`

**Comment by kiepczi** [comment]
An initial module for parsing delta files has been added to branch `issue_4`. The code was tested against values returned by the current `pyani v0.3`, and all tests pass.

I am currently working on writing a Snakemake workflow to parse `delta` files for ANIm calculations.

**Comment by widdowquinn** [comment]
> I am currently working on writing a Snakemake workflow to parse `delta` files for ANIm calculations.

We should talk in more detail about this as I'm not sure what the design intent is, here. We need to break down what we're doing into self-contained tasks, and the task at hand here is to parse a delta file into an internal representation. That doesn't need snakemake, but could reuse code from the `pyani` repository (specifically https://github.com/widdowquinn/pyani/blob/master/pyani/nucmer.py)

**Comment by kiepczi** [comment]
Hi @widdowquinn, 

Yes, you are right. I'm currently reusing the same code from pyani to calculate ANI values. However, the last time we spoke about it, you suggested using pyroject.toml to create an executable script for this so we can use snakemake rules to run the script. Am I missing something?

I agree that this might be unnecessary because parsing delta files does not take a lot of time. Maybe if we wish to use pyani-plus on bigger datasets, it would make a difference, but that would need to be tested, I guess.

**Comment by widdowquinn** [comment]
> the last time we spoke about it, you suggested using pyroject.toml to create an executable script for this so we can use snakemake rules to run the script. Am I missing something?

I don't think you're missing anything, but there are two steps being munged together in what may not be the best order. We need to build up the software in steps, something like:

- workflow to run the necessary comparisons
- module with a parser to correctly parse the comparison output (A)
- module to calculate the distance/comparison value from the parsed output (B)
- module to add a comparison result to the database [assuming the database is built/available] (C)
- script combining A, B, and C
- workflow that executes this script at the appropriate time

Going in this order means we can build on solid foundations, knowing the earlier steps in the process are working and unlikely to require us to go back and rewrite lots of code. The workflow is arguably the last thing (in sequence) that we need to write in this process. We _could_ write it at any point, but if we write it before the script is available we may not know enough about how to call the script, and end up rewriting it anyway, which is lost effort.

You don't build a stone arch without a [keystone](https://en.wikipedia.org/wiki/Keystone_(architecture)), but you can't put the keystone in first - it's about the order in which we develop things, not that the workflow is unnecessary.

**Comment by widdowquinn** [comment]
> Yes, you are right. I'm currently reusing the same code from pyani to calculate ANI values. 

I can see how you might have got that intent from my badly-worded issue name/title. I've changed the name of the issue to restrict the scope to something more achievable and appropriate.

**Comment by kiepczi** [comment]
So, I have been working on this in the `issue_4` branch.

As of now, I have written a module that calculates the values for the aniM method. In fact, I copied the module from the previous `pyani` codebase and reformatted it based on the recommendations of `ruff`.

I have also written tests to verify the values against what we have already agreed upon with the current working version of `pyani`.

We discussed this in today's meeting, but I'll note it here so we don't forget: `ruff` does not like the use of the assert statement in the code, raising an [S101](https://docs.astral.sh/ruff/rules/assert/) error. We agreed that this is not problematic for `pytest`, and we want to continue using `assert` for this purpose. I have added a few lines in `pyproject.toml` to ignore the S101 error for all tests.

**Comment by widdowquinn** [comment]
Thanks @kiepczi - just to add a note from the discussion, we talked about ignoring errors for files matching `tests/test_*.py` to the `ruff` config in `pyproject.toml` so that we could avoid the `S101` warnings.

**Comment by kiepczi** [comment]
At today's meeting, we discussed improving the code by removing repetitive lines and replacing them with a function.

For example, our previous lines for calculating the lengths of aligned bases in the query and reference sequences without overlaps were as follows:

```python
    for seq_id in regions_qry:
        qry_tree = intervaltree.IntervalTree.from_tuples(regions_qry[seq_id])
        qry_tree.merge_overlaps(strict=False)
        for interval in qry_tree:
            qaln_length += interval.end - interval.begin + 1

    for seq_id in regions_ref:
        ref_tree = intervaltree.IntervalTree.from_tuples(regions_ref[seq_id])
        ref_tree.merge_overlaps(strict=False)
        for interval in ref_tree:
            raln_length += interval.end - interval.begin + 1
```
I have now written a separate function that can be called within the `parse_delta` function for both query and reference sequences:

```python
def get_aln_length(aln_regions: dict) -> int:
    """Return the number of aligned bases (without overlaps).

    :param aln_regions: dict of aligned regions
    """
    aln_length = 0
    for seq_id in aln_regions:
        tree = intervaltree.IntervalTree.from_tuples(aln_regions[seq_id])
        tree.merge_overlaps(strict=False)
        for interval in tree:
            aln_length += interval.end - interval.begin + 1

    return aln_length
```

**Comment by kiepczi** [comment]
I should probably mention that I also checked how this change affected the runtime of a single parse. The time did not decrease, but CPU usage dropped slightly, which I suppose is a plus?

eg. previous code:
```(pyani_plus_dev) akiepas@macstudio methods % time python old_method_anim.py
python old_method_anim.py  0.04s user 0.01s system 83% cpu 0.062 total
```

after modifications:
```(pyani_plus_dev) akiepas@macstudio methods % time python method_anim.py    
python method_anim.py  0.04s user 0.01s system 86% cpu 0.056 total
```

**Comment by peterjc** [comment]
Changes of that level are more or less noise - but we can say there was no unexpected performance regression.

**Comment by widdowquinn** [comment]
Hi @kiepczi - just to be sure, you want me to take a look at the `issue_4` branch at commit e93f1dc?

**Comment by kiepczi** [comment]
Hi @widdowquinn - yes. 

I want to see if I understood what we talked about yesterday, so I don't waste time on something that is wrong or pointless. 
If that's the idea how we want to catch the information by parsing `delta` files, I will try to restructure this code by separating it in different scripts, as you showed me yesterday based on old `pyani`.

**Comment by widdowquinn** [comment]
I've merged `main` into `issue_4` so that everything's up to date, and am working forward from that point.

**Comment by widdowquinn** [comment]
I've been through the code in that commit, and made some changes. I hope that they're either self explanatory or adequately documented. You can see the differences (including the merged files from `main`) at the link below:

- https://github.com/pyani-plus/pyani-plus/compare/6b53037..e93f1dc

where the new code is on the left. My changes don't kick in until you get to `method_anim.py`. I'll try to explain my rationale in order of occurrence in each file, though there is some cross-linking.

- I hope the copyright change is obvious ;)
- I've explained why a NamedTuple rather than a data class (and you made the right choice!) for `ComparisonResult`
- I've added a TODO here because the comparison also needs coverage information
- I've modified some attribute names for clarity, as "length" could be total length or aligned length, or aligned base count if not clarified
- I changed the name of `get_aln_length()` to `get_aligned_bases_count()` for the same reason
- I expanded some variable names for clarity
- In `parse_delta` we didn't need to declare initial values for variables that weren't being modified (e.g. `current_ref`, etc.)
- I updated some of the comments where they hadn't followed the type change
- I changed `weighted_identical_bases` from a list to a count, to match the other changes
- I moved the calculation of the aligned base counts into the return statement. If you think it's clearer the other way I'm happy for you to restore it.
- I changed the name of `update_comparison_results()` as I thought `collect_results_directory()` was a bit more specific, though I'm open to better names
- I added a TODO comment to add a test for this function code
- I modified variable names to be a bit more explicit

The big change here compared to the old `pyani` - which I added in a comment at the appropriate point - is that you've now got the code that parses a complete run's output directory into the core module code (the equivalent of `anim.py` or `nucmer.py` in the original codebase), rather than in the CLI (`subcmd_anim.py` in `pyani`). I don't mind that, but we should probably have a more comprehensive meeting about the overall architecture of what we're doing, and solidify it with a reference in the `design-docs` repository.

**Comment by kiepczi** [comment]
Thanks, @widdowquinn! That is very helpful.

I've been working on this today, and you can find the progress under the [c1be097 commit](https://github.com/pyani-plus/pyani-plus/commit/c1be09717d64a8a1d373502e8db057a0ee6fa40f).

I’ve added genome coverage calculations for both the query and the reference, as well as Hadamard calculations. Based on the previous Hadamard calculations in `pyANI` ([available here](https://github.com/widdowquinn/pyani/blob/69efdf6efcb154667a39c0f57839a7fb51e70f70/pyani/pyani_tools.py#L170)), it’s calculated as `genome identity * genome coverage`. However, if this is incorrect, I apologise and am open to corrections.

For the genome coverage calculations, I needed to get the total lengths of genomes, which I achieved with the function below. I know that genome lengths will be needed to calculate values for other methods (e.g., `dnadiff`). It might be worth having a separate script that can be called when needed, but (as @widdowquinn briefly mentioned today), we need to talk about how we would like to structure `pyani_plus`. I'm just putting this here so we don't forget when the next meeting arrives.
```python
def get_genome_length(filename: Path) -> int:
    """Return total length of all sequences in a FASTA file.
    :param filename:  path to FASTA file
    """
    with Path.open(filename) as ifh:
        return sum([len(record) for record in SeqIO.parse(ifh, "fasta")])
```


Please keep in mind that I still need to write tests to ensure the code is working as expected.

**Comment by peterjc** [comment]
Sounds good. It probably doesn't matter, but we can speed up that ``get_genome_length`` function using a different bit of Biopython.

**Comment by kiepczi** [comment]
That would be great. Unfortunately that's the only way I know, so I'm happy to suggestions ;)
Do you have something specific in mind. It would be great if you could point me in the right direction.

**Comment by widdowquinn** [comment]
Thinking about it, it's probably a good idea for us to take a look at the database schema on Tuesday and bring it into the codebase as early as we can. It would be wasteful to write a lot of placeholder code we know we need to remove later, which is where I think this discussion is potentially leading.

**Comment by kiepczi** [comment]
Closing this issue as it was solved in #100 and #107.
